This pipeline is designed to process RNA-Seq data generated by total RNA-Seq strategy. One advantage of total RNA-Seq 
strategy is it can capture and sequence both linear and circular mRNA isoforms at a single run. circRNA is not 
considered in most existing gene expression quantification tools and pipelines which focused on polyA enriched 
RNA-Seq data. To solve this problem, we designed this total RNA-Seq analysis pipeline. It identifies circRNAs 
from total RNA-Seq first. Then, RNA-Seq reads will be distributed to linear and circular mRNA isoforms to quantify
their expression.


## Quick Start

1. Install [Nextflow](https://www.nextflow.io/docs/latest/getstarted.html#installation) (>=21.04.1)
2. Install [docker](https://docs.docker.com/engine/install/) for pipeline reproducibility
3. Prepare a case ID file (see [documentation](https://github.com/bzhanglab/nf-rnaseq/wiki) for file format) 
4. Run a test with 2 samples on local server.

```console
nextflow run bzhanglab/nf-rnaseq -r main -profile docker \
   --case_id /path/to/my_case_id.txt \
   --start 1 --end 2 --run_version rnaseq-2021-01-01
```
5. Run with AWS batch on AWS cloud, see [here](https://www.nextflow.io/docs/latest/awscloud.html) for 
   proper configuration of your AWS environment.

```console
nextflow run bzhanglab/nf-rnaseq -r main -profile awsbatch \
   --start 1 --end 2 \
   --case_id /path/to/my_case_id.txt \
   -bucket-dir s3://zhanglab-nextflow-workdir/workdir/2021-01-01 \
   --outdir /path/to/rnaseq-results \
   --run_version rnaseq-2021-01-01
```

<!-- In the above examples, if you want the results to be stored locally on your 
launch host, just provide `--run_version xyz`, and the results will be stored
at `results/xyz` under your launch directory. 

To run with awsbatch, you must specify an s3 path as `outdir`, e.g.
`--outdir s3://my-bucket/my-folder`.  In this case, results will be 
stored under `s3://my-bucket/my-folder/run_version`. -->


<!-- ## Inputs

Required:

* CPTAC3 catalog file from https://github.com/ding-lab/CPTAC3.catalog/blob/master/CPTAC3.Catalog.dat (S3 path specified in the config file)

User inputs:

* Case ids file called `case_id.txt` contains all the case ids to be processed in a column.
  This file can be in the local directory where nextflow is launched or in S3.
   -->
